extmod/moducryptolib.c

```diff
@@ -119,6 +144,33 @@ STATIC void aes_process_cbc_impl(AES_CTX_IMPL *ctx, const uint8_t *in, uint8_t *
         AES_cbc_decrypt(ctx, in, out, in_len);
     }
 }
+
+#if MICROPY_PY_UCRYPTOLIB_CTR
+// axTLS doesn't have CTR support out of the box. This is based on `mbedtls_aes_crypt_ctr`.
+STATIC void aes_process_ctr_impl(AES_CTX_IMPL *ctx, const uint8_t *in, uint8_t *out, size_t in_len, struct ctr_params *ctr_params) {
+    size_t n = ctr_params->nc_off;
+    uint8_t *const nonce_counter = ctx->iv;
+
+    while (in_len--) {
+        if (n == 0) {
+            aes_process_ecb_impl(ctx, nonce_counter, ctr_params->stream_block, true);
+
+            for (int i = 16; i > 0; i--) {
+                if (++nonce_counter[i - 1] != 0) {
+                    break;
+                }
+            }
+        }
+        const int c = *in++;
+        *out++ = (unsigned char)(c ^ ctr_params->stream_block[n]);
```

Comment:
Either useless, or incorrect cast. out is uint8_t*.

Reply:
Technically `c` is a an `int`, so the expression `c ^ ...` is promoted to an `int` as well, hence the cast.
Anyway this is copied from `mbedtls_aes_crypt_ctr` and I attempted to keep the code as identical as possible.

Reply:
> Anyway this is copied from mbedtls_aes_crypt_ctr

If it's copied, there's a potential copyright problem, and potential licensing problem (mbedtls comes with a different license than MicroPython).

Reply:
If this function is indeed copied in anyway then that's a big red flag...

Reply:
Well, we can count on it to work perfectly. I trust mbedtls's impl of AES, after all.

But I admit I didn't have the licensing in mind, so I'll rewrite this myself and fix all formatting / naming issues.